A variety of changes related to HUBOL

ShinyTillDawn opened the pull request:
− Fixed a typo.
− Added an updated `moveeffectoptions` [[List]] for newly−created HUBOL ROMs.
− Fixed the number of moves to recognize.
− Adding support for the new, 24x24 type chart.

Everything's been tested, and everything seems fine to me.

haven1433 commented on src/HexManiac.Core/Models/HardcodeTablesModel.cs:
```diff
@@ -349,6 +349,74 @@ public const string
          newList.Add("Evsn Up 3");
          SetList(new NoDataChangeDeltaModel(), "zeffects", newList, null, StoredList.GenerateHash(newList));
+		 
+	 // Move effects in HUBOL
```
use spaces instead of tabs in order to make the whitespace consistent

haven1433 commented:
@ShinyTillDawn 
Some hubol users have had trouble with editing the type chart. Hubol may be inserting a replacement chart, leaving HMA to point at an unused original chart. Can you investigate?

ShinyTillDawn commented:
I tried searching for bytes that would indicate the presence of a new type chart a few months ago, but to no avail. I can try again and fix the spacing.

haven1433 commented:
Looks pretty Good! I'll request that you use ".5x" instead of "½x" for effectiveness 5, so that it's easier to type when editing the table. Or if ".5x" causes trouble because of the `.`, maybe "halfx"?

ShinyTillDawn commented:
I changed the [[List]] item's name internally and fixed the spacing. It's gonna be hard implementing the type chart support without making a mistake...

haven1433 commented on src/HexManiac.Core/Models/HardcodeTablesModel.cs:
```diff
@@ -198,26 +198,70 @@ public const string
+            newList[207] = "RaiseAllStatsPrimary";
+            newList[213] = "Stat Swap or Split";
+            for(var k = 0; k < 4; k++) newList.Add(null); // 214 to 217 are unused.
             newList.Add("Me First");
```
Go ahead and use `newList.AddRange(...)` here like we talked about, to make it easier to read the list and add new elements. Something like...

```cs
newList.AddRange(new [] {
  "Me First",
  "Eat Berry",
  "Natural Gift",
  ...
  "Last Resort or Sky Drop",
  "Damage Set Terrain",
  "Teatime",
});
```

ShinyTillDawn replied:
Previously, this function didn't use `AddRange()` at all. Also, I can't really use `AddRange()` here because I'm _editing_ existing items in the list. Using that list method would require me to rebuild the list from the ground up, which isn't necessary. After `newList[213]`, I could use `AddRange()` as that's the point where new elements are added to the table.

haven1433 commented on src/HexManiac.Core/Models/HardcodeTablesModel.cs:
```diff
@@ -246,6 +290,20 @@ public const string
+            var newMultipliers = new List<string>();
+            newMultipliers.Add("1x");
+            newMultipliers.Add("0x");
+            while (newMultipliers.Count <= 5)
```
Probably the clearest way to do this is to make all the nulls first, and then insert into the indexes that aren't null.
```cs
var newMultipliers = new string[21]; // 0 to 20
newMultipliers[0] = "1x";
newMultipliers[1] = "0x";
newMultipliers[5] = "0.5x";
newMultipliers[20] = "2x";
SetList(new NoDataChangeDeltaModel(), "effectiveness", newMultipliers, null, StoredList.GenerateHash(newMultipliers));
```

ShinyTillDawn replied:
That works even better!

haven1433 commented on src/HexManiac.Core/Models/HardcodeTablesModel.cs:
```diff
@@ -349,6 +407,78 @@ public const string
          newList.Add("Evsn Up 3");
          SetList(new NoDataChangeDeltaModel(), "zeffects", newList, null, StoredList.GenerateHash(newList));
+         
+         // Move effects in HUBOL
```
I think this commented functionality is handled elsewhere... please remove it before the review is published.

ShinyTillDawn replied:
This was the first attempt at me trying to add the updated move effects, but it's no longer needed after pushing this commit. Thus, I don't think that commented functionality is handled elsewhere.

haven1433 commented on src/HexManiac.Core/Models/HardcodeTablesModel.cs:
```diff
@@ -506,6 +636,14 @@ public const string
+         // type chart
+         if (name == "data.pokemon.type.chart") {
+            source = 0x145BB74;
+            format = "[Norm.effectiveness Fight.effectiveness Fly.effectiveness Poison.effectiveness Grd.effectiveness Rock.effectiveness Bug.effectiveness Ghost.effectiveness Steel.effectiveness param9.effectiveness Fire.effectiveness Water.effectiveness Grass.effectiveness Elec.effectiveness Psych.effectiveness Ice.effectiveness Drag.effectiveness Dark.effectiveness param18.effectiveness param1 ...
```
To make this easier to read, we might construct it up from a list of the types first, and then add the `.` and `effectiveness` annotation. Will also help guard against typos. Maybe something like:
```cs
format = "Norm Fight Fly Poison Grd Rock Bug Ghost Steel t9 Fire Water Grass Elec Psych Ice Drag Dark t18 t19 t20 t21 t22 Fairy";
format = " ".Join(format.Split(' ').Select(type => $"{type}.effectiveness"));
format = $"[{format}]data.pokemon.type.names"
```

ShinyTillDawn replied:
This would definitely be more optimized.

haven1433 reviewed (changes requested):
Looking pretty good! Just a few suggestions to make the code a bit easier to read and work with.

haven1433 commented:
It's looking pretty good to me. Mark it as ready for review (not draft) when you're done with the functionality, then I'll do a final pass review before accepting the PR.
